Allow index.html to be served from file protocol

swim opened the pull request:
Hey Kyle, hope you're well. 

recently we've had a situation where Gatsby had to be built and shipped off to a closed system. There are extreme security requirements around this system and therefore no web server could be used to serve Gatsby. While setting pathPrefix eliminated the majority of our issues we still had a problem serving the index of each page / the index.html from each subfolder. Named pages, such as 404.html worked fine.

KyleAMathews commented:
I am well :-) Hope things are nice with you too!

Patch looks good! Hadn't thought about this use case but seems easy enough to support. Could you add a comment above the if statement explaining why this would happen?

swim commented:
Should have dropped this in my first pr; thank you so much for Gatsby! It's products like this that makes programming enjoyable =).

Updated the PR to use the more appropriate React router method matchPath. Added a comment on why the condition should trigger.

KyleAMathews commented:
Sweet, thanks!

jayliu50 commented:
@swim sorry to hijack this thread, but you mentioned that you were getting Gatsby to work through the file protocol. I have similar requirements. What do you set the pathPrefix to? Do you mind sharing tips? Any/all would be greatly appreciated!

swim commented:
Hey @jayliu50 
no problem, I'm sorry about my slow response. We set the pathPrefix with a token value which is then replaced by an additional post process. The pathPrefix needs to be the location on the target system; we used a token because this could change. Below I've provided an example of how you can modify the path through gatsby-node.js. Furthermore, you could override the default static-entry.js and have control over the final build steps.

`
var gatsbyConfig = require('./gatsby-config');

exports.modifyWebpackConfig = ({ config, stage }) => {
  var pathPrefix = gatsbyConfig.pathPrefix;

  if (stage == 'build-css') {
    config._config.output.publicPath = pathPrefix + '/'
  }

  if (stage == 'build-javascript') {
    config._config.output.publicPath = pathPrefix + '/'
  }

  if (stage === 'build-html') {
    config._config.entry = {
      main: [require.resolve('babel-polyfill'), path.resolve(__dirname, 'static-entry.js')]
    }
  }

  return config;
};
`

Hope this helps,
